### ISSUE TITLE
Glslang HLSL produces seemingly incorrect OpImageRead on RWTexture2D<float>

### ISSUE THREAD

**Issue by hugoam** [KhronosGroup/glslang#2613]
When compiling the following HLSL compute shader, glslang emits an OpImageRead which gets subsequently refused by `spirv-opt` with the following error:
```
Error: Expected Result Type to have 4 components
  %119 = OpImageRead %float %107 %106
```

```
cbuffer UniformBlock
{
 float4 u_inputRTSize;
};

float4 imageLoad(RWTexture2D<float> _image, int2 _uv) { return _image[_uv].xxxx; }
void imageStore(RWTexture2D<float> _image, int2 _uv, float4 _value) { _image[_uv] = _value.x; }

[[spv::format_r32f]] RWTexture2D<float> s_texOcclusionDepth : register(u1); ;

[numthreads(16, 16, 1)]

void main( int3 gl_GlobalInvocationID : SV_DispatchThreadID )
{
int2 coord = int2(gl_GlobalInvocationID.xy);
float maxDepth = imageLoad(s_texOcclusionDepth, int2(u_inputRTSize.zw * coord.xy ) ).x;
imageStore(s_texOcclusionDepth, coord, float4(maxDepth,0,0,1) );
}
```

http://shader-playground.timjones.io/ae167b61c24705aee8ab7bbfbc72928e

Is this a glslang bug, or is something else the issue here ?

This seems to come from this piece of validation code in `spirv-opt`:
```
  // Vulkan requires the result to be a 4-element int or float
  // vector.
  if (spvIsVulkanEnv(target_env)) {
    if (_.GetDimension(actual_result_type) != 4) {
      return _.diag(SPV_ERROR_INVALID_DATA, inst)
             << "Expected " << GetActualResultTypeStr(opcode)
             << " to have 4 components";
    }
  }  // Check OpenCL below, after we get the image info.
```

We did specify something like `shader->setEnvClient(glslang::EShClientVulkan, EShTargetVulkan_1_0);`, so I suppose glslang is emitting some ill formed Vulkan SPIRV instructions here, right ?

**Comment by greg-lunarg** [comment]
Yes. Sorry. This is a bug and is already documented in #2486.

Vulkan indeed requires 4 elements from OpImageRead. glslang should give an error and should not produce SPIR-V for this shader.

I will close this in deference to #2486 and will endeavor to fix this problem soon.

**Comment by hugoam** [comment]
>glslang should give an error and should not produce SPIR-V for this shader.

Hmm I did not expect that the source shader could be construed as wrong. Do you mean we can only use RWTexture2D<float4> when targeting Vulkan ? I did not expect that. This is to accommodate OpImageRead, but does OpImageStore always mandates 4 components in the same fashion ?

Also while trying to find more information on this, I sadly can't find the document that specifies that, if you happen to know where to find it that would help me understand this :)

**Comment by greg-lunarg** [comment]
Sorry. Yes. Thank you for the correction.

The correct behavior here is, when targeting Vulkan, for glslang to always generate an OpImageRead whose return type is a 4-component vector of the sampled type of the image. In this case, the type should be vec4.

Here is the spec: https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html

Here is the quote from the spec:

`The Result Type operand of any OpImageRead or OpImageSparseRead instruction must be a vector of four components.`

For OpImageWrite, the texel does not have to be 4-component vector, although it does need to have as many components as the view format. Here are two quotes from the spec related to this:

`If a VkImageView is accessed using OpImageWrite as a result of this command, then the Type of the Texel operand of that instruction must have at least as many components as the image view’s format.`

`If a VkBufferView is accessed using OpImageWrite as a result of this command, then the Type of the Texel operand of that instruction must have at least as many components as the buffer view’s format.`

I will keep this closed in deference to #2486
